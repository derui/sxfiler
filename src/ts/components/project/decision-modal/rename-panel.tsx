import { h } from "preact";
import { useEffect, useRef, useState } from "preact/hooks";

type Handler = (newName: string) => void;

export type Props = {
  selected: boolean;
  onUpdated: Handler;
  itemName: string;
};

const handleChange = (cb: (input: string) => void) => (ev: preact.JSX.TargetedEvent<HTMLInputElement, Event>) => {
  cb(ev.currentTarget.value || "");
};

export const Component: preact.FunctionComponent<Props> = ({ selected, itemName, onUpdated }) => {
  const [state, setState] = useState(itemName);
  const refInput = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (refInput.current) {
      refInput.current.focus();
    }
  }, []);

  return (
    <div class="decision-modal__panel" data-testid="decisionModal-rename" aria-selected={selected}>
      <p class="decision-modal__label" data-testid="decisionModal-renameLabel">
        Rename
      </p>
      <label>
        <input
          class="decision-modal__input"
          data-testid="decisionModal-input"
          type="text"
          placeholder="New name of the node"
          value={state}
          onInput={handleChange((v) => {
            setState(v);
            onUpdated(v);
          })}
          ref={refInput}
        />
      </label>
    </div>
  );
};
