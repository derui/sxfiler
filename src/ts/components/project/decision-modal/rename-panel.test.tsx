import { h } from "preact";
import { render } from "preact-render-to-string";
import { Component } from "./rename-panel";
import { render as mount, fireEvent } from "@testing-library/preact";

describe("Project", () => {
  describe("Rename decision", () => {
    it("render with rename decision", () => {
      const handler = () => {};
      const tree = render(<Component selected={false} onUpdated={handler} itemName={"node"} />);

      expect(tree).toMatchSnapshot();
    });

    it("change style when selected", () => {
      const handler = () => {};
      const tree = render(<Component selected={true} onUpdated={handler} itemName={"node"} />);

      expect(tree).toMatchSnapshot();
    });

    it("do not render with other decision", () => {
      const handler = () => {};
      const tree = render(<Component selected={false} onUpdated={handler} itemName={"node"} />);

      expect(tree).toMatchSnapshot();
    });

    it("call handler when the new name in the component is changed", (done) => {
      const dom = document.createElement("div");
      const handler = (newName: string) => {
        expect(newName).toEqual("foo");
        done();
      };

      const per = mount(<Component selected={true} onUpdated={handler} itemName="bar" />, { baseElement: dom });
      const v = per.getByTestId("decisionModal-input") as HTMLInputElement;
      v.value = "foo";
      fireEvent.input(v, new Event("input"));
    });
  });
});
