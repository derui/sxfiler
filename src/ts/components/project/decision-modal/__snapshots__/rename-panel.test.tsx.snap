// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Rename decision change style when selected 1`] = `"<div class=\\"decision-modal__panel\\" data-testid=\\"decisionModal-rename\\" aria-selected><p class=\\"decision-modal__label\\" data-testid=\\"decisionModal-renameLabel\\">Rename</p><label><input class=\\"decision-modal__input\\" data-testid=\\"decisionModal-input\\" type=\\"text\\" placeholder=\\"New name of the node\\" value=\\"node\\" /></label></div>"`;

exports[`Project Rename decision do not render with other decision 1`] = `"<div class=\\"decision-modal__panel\\" data-testid=\\"decisionModal-rename\\"><p class=\\"decision-modal__label\\" data-testid=\\"decisionModal-renameLabel\\">Rename</p><label><input class=\\"decision-modal__input\\" data-testid=\\"decisionModal-input\\" type=\\"text\\" placeholder=\\"New name of the node\\" value=\\"node\\" /></label></div>"`;

exports[`Project Rename decision render with rename decision 1`] = `"<div class=\\"decision-modal__panel\\" data-testid=\\"decisionModal-rename\\"><p class=\\"decision-modal__label\\" data-testid=\\"decisionModal-renameLabel\\">Rename</p><label><input class=\\"decision-modal__input\\" data-testid=\\"decisionModal-input\\" type=\\"text\\" placeholder=\\"New name of the node\\" value=\\"node\\" /></label></div>"`;
